bloomFilter_test.go

```diff
@@ -1,85 +1,109 @@
+		item := randomBytes(10)
 
-// Basic init test
-func (s *MySuite) TestNew(c *C)  {
-	bf := New(1024, 3)
+		bf.Add(item)
 
-	c.Assert(bf.k, Equals, uint(3))
-	c.Assert(bf.n, Equals, uint(0))
-	c.Assert(bf.m, Equals, uint(1024))
-}
+		if bf.Test(item) != true {
+			t.Errorf("'%q' not found", item)
+		}
 
-// Test add to set
-func (s *MySuite) TestAdd(c *C)  {
-	bf := New(1024, 3)
+		// Now lets create some items that don't exist
+		item2 := append(item, randomBytes(10)...)
 
-	bf.Add([]byte("hello"))
+		// Test that item does NOT exist
+		if bf.Test(item2) == true {
+			t.Errorf("'%q' should not be found", item2)
+		}
+	}
+}
 
-	c.Assert(bf.n, Equals, uint(1))
+func BenchmarkAdd(b *testing.B) {
+	bf := New(memberSize, DefaultHashFunctions)
+	for i := 0; i < b.N; i++ {
+		bf.Add(randomBytes(10))
+	}
 }
 
-// Test items if do not exist into set
-func (s *MySuite) TestIfNotExist(c *C)  {
-	bf := New(1024, 3)
+func BenchmarkTest(b *testing.B) {
+	bf := New(memberSize, DefaultHashFunctions)
+	for i := 0; i < b.N; i++ {
+		bf.Add(randomBytes(10))
+	}
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		bf.Test(randomBytes(10))
+	}
+}
 
-	bf.Add([]byte("hello"))
-	bf.Add([]byte("world"))
-	bf.Add([]byte("hi"))
+// Compare with a binary sort to make sure we're in the same ballpark
+func BenchmarkBinarySearch(b *testing.B) {
+	var strings []string
+	for i := 0; i < sampleSize; i++ {
+		item := randomString()
+		strings = append(strings, item)
+	}
 
-	c.Assert(bf.Test([]byte("H1")), Equals, false)
-	c.Assert(bf.Test([]byte("World")), Equals, false)
-	c.Assert(bf.Test([]byte("hell0")), Equals, false)
-}
+	// Sort by byte order
+	sort.Strings(strings)
 
-// Test items if items may exist into set
-func (s *MySuite) TestIfMayExist(c *C)  {
-	bf := New(6, 3)
-
-	bf.Add([]byte("hello"))
-	bf.Add([]byte("world"))
-	bf.Add([]byte("sir"))
-	bf.Add([]byte("madam"))
-	bf.Add([]byte("io"))
-
-	c.Assert(bf.Test([]byte("hello")), Equals, true)
-	c.Assert(bf.Test([]byte("world")), Equals, true)
-	// False negative
-	c.Assert(bf.Test([]byte("hi")), Equals, false)
-}
+	b.ResetTimer()
 
-func (s *MySuite) BenchmarkAdd(c *C) {
-	bf := New(1024, 3)
-	for i := 0; i < c.N; i++ {
-		// Logic to benchmark
-		bf.Add([]byte(string(i)))
+	for i := 0; i < b.N; i++ {
+		item := randomString()
+		use(sort.SearchStrings(strings, item))
 	}
 }
 
-func (s *MySuite) BenchmarkTest(c *C) {
-	bf := New(1024, 3)
-	for i := 0; i < c.N; i++ {
-		// Logic to benchmark
-		bf.Add([]byte(string(i)))
+func BenchmarkHashFunctions(b *testing.B) {
+	item := []byte("hello world, how are you doing?")
+
+	hashFns := map[string]hash.Hash64{
+		"murmur3": murmur3.New64(),
+		"fnv64":   fnv.New64(),
+		"fnv64a":  fnv.New64a(),
 	}
-	c.ResetTimer()
-	for i := 0; i < c.N; i++ {
-		// Logic to benchmark
-		bf.Test([]byte(string(i)))
+
+	for name, hashFunc := range hashFns {
+		b.Run(name, func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				hashFunc.Write(item)
+				use(hashFunc.Sum64())
```

Comment:
What does `use` do?

Reply:
I added `use()` to keep the compiler from optimizing the return out of code.

Reply:
Shouldn't -gcflags=-N do the same or not?

Reply:
I'm not sure, I wasn't aware that `gcflags` affected code execution paths outside of disabling inlining of function invocations and such.